### ISSUE TITLE
Store energies as float64 in HDF5 file

### ISSUE THREAD

**Issue by jchodera** [openmm/spice-dataset#55]
The downloader currently stores the formation energies in the HDF5 [using `float32`](https://github.com/openmm/spice-dataset/blob/main/downloader/downloader.py#L115), but this leads to the introduction of [errors that are a good fraction of 1 kcal/mol into the energies](https://github.com/openmm/spice-dataset/issues/54).

To remedy this, we should
1. Change the [HDF5 downloader](https://github.com/openmm/spice-dataset/blob/main/downloader/downloader.py#L115) to use `float64`
2. Rebuild and re-release the HDF5 file
3. Retrain the SPICE models in the [preprint]() using the updated HDF5 file

**Comment by peastman** [comment]
> Retrain the SPICE models in the [preprint](https://github.com/openmm/spice-dataset/issues/55) using the updated HDF5 file

As I noted in https://github.com/openmm/spice-dataset/issues/54#issuecomment-1314073728, that would make no difference.  Also, training in fp64 would be extremely slow, and the resulting model would be too slow to be usable on most GPUs.

**Comment by giadefa** [comment]
usually, the dataset is converted to float32 in the data loader. Float64 is
never used for training.

**Comment by jchodera** [comment]
> Also, training in fp64 would be extremely slow, and the resulting model would be too slow to be usable on most GPUs.

We are not talking about training in `float32`---just storing the data in `float64`.

The dataset is converted to `float32` in the [data loader](https://github.com/torchmd/torchmd-net/blob/main/torchmdnet/datasets/spice.py#L231-L233). 

If the loss function is not sensitive to errors of ~0.25 kcal/mol, this will cause large errors conformational energies. 

We could split the loss function into two components: Measuring the error in the total formation energy (which will be only crudely accurate) and measuring the error in relative conformer energies (relative to one conformation or the average over conformations).

> But this will have no effect at all on any of our experiments. We should be so lucky as to have a model that could match the training data to 1e-7! The training error is nowhere close to the point where errors at that level make a difference.

The [paper reports the median absolute error as 2.912 kJ/mol](https://arxiv.org/abs/2209.10702), which is 0.70 kcal/mol, and the error in our `float32` total energies can be as large as 0.25 kcal/mol (possibly larger). I'm not quite understanding why we think that eliminating errors in the training data on the order of magnitude as the test set error would not impact the resulting model performance.

**Comment by peastman** [comment]
> If the loss function is not sensitive to errors of ~0.25 kcal/mol, this will cause large errors conformational energies.

We're talking about far smaller errors than that.  I just ran an analysis on the dataset.  The median formation energy is -13428.82 kJ/mol, and the very largest formation energy of any conformation is -35704.156.  Stored in single precision (machine epsilon is 5e-8), the errors due to lack of precision are about 0.001 kJ/mol.  That's three orders of magnitude smaller than the errors in fitting the model.

**Comment by peastman** [comment]
Can we clarify which fields should be in double?  As noted above, the premise of this issue is incorrect: errors in formation energies are only about 0.001 kJ/mol.  Are there any applications for which that's a problem?  The total energies do have significant errors, because they're much larger, so that field should be double.  Are there any others?

**Comment by pavankum** [comment]
I think `dft_total_gradient` is fine since 1e-08 hartree/bohr is of the order of 1e-03 kJ/mol/nm. I don't see the need for other fields.

**Comment by jchodera** [comment]
> the premise of this issue is incorrect: errors in formation energies are only about 0.001 kJ/mol

How do you reconcile this with the [reported error of 0.25 kcal/mol](https://github.com/openmm/spice-dataset/issues/54#issue-1444455213)?

**Comment by jchodera** [comment]
If energy errors are 0.25 kcal/mol, this seems like an urgent issue. If they're 0.001 kcal/mol, it seems much less urgent.
I'm just not certain how to reconcile the reported errors from #54 (0.25 kcal/mol) with your analysis (0.001 kcal/mol).
Perhaps there is a _separate_ issue causing #54?

**Comment by peastman** [comment]
That post referred to the energy returned by `qcng.compute()`, which is the total energy.  It tends to be about 1000x larger than the formation energy, so the truncation error is correspondingly larger.

There's probably no harm in using double for formation energy too, though.  Energies take up so much less space than a lot of other fields, it will have a negligible impact on file size.

**Comment by jchodera** [comment]
Thanks for clarifying this! If I understand correctly, this means all practical uses (including the ML models) use the formation energy, where the error is ~ 0.001 kcal/mol and irrelevant. But for completeness, to ensure error is small for total energy too, we should rebuild the HDF5 file with at least the total energy in `float64`. Is that right?

**Comment by peastman** [comment]
Correct.  I just opened a PR to do that.  I used double for formation energies too, since there's not much reason not to.  This only adds 8 bytes per conformation, or about 9 MB for the whole dataset, which is currently over 10 GB.
